routing/payment_session.go

```diff
@@ -60,6 +60,7 @@ func (p *paymentSession) RequestRoute(payment *LightningPayment,
 	// If we have a pre-built route, use that directly.
 	case p.preBuiltRoute != nil && !p.preBuiltRouteTried:
 		p.preBuiltRouteTried = true
+		p.preBuiltRoute.AddFinalHopRecords(payment.FinalDestRecords)
```

Comment:
This brings up the discussion about how to represent the mpp fields internally. In my understanding, `FinalDestRecords` are untyped and only meant for application-specific fields from the custom range.

I can understand that this may be easy for now, but don't we eventually want to add proper fields for mpp and amp to the hop structure, just as `amt` is a proper field? And then convert that into the tlv stream only in PackHopPayload. This is then also where the tlv id is known for those fields.

Reply:
`FinalDestRecords` is indeed untyped at this stage, but the intention was that a higher layer would have already serialized the complete tlv payload at that point, using richer types at an earlier stage of the process. 

As far as adding additional fields to the `Hop` payload, if they're all in-line, rather than an optional pointer to a richer struct, we'll end up with fields that are mostly not used (depending on which one becomes the dominant payment type) sitting there in memory FWIW.